### PR TITLE
CLDR-15034 kbd: Keyboard 3.0 Spec

### PR DESCRIPTION
**Pull request by srl295**
- Update to the TR35 part 7 Keyboard spec, to harmonize with PRI 476 (techpreview)

Co-authored-by: Marc Durdin <marc@durdin.net>
(cherry picked from commit f3d2f68b2843a2e1939191a7ecd18ab974374d0a)

CLDR-15034

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

### CODE REVIEW

**Comment by srl295** [comment]
Review Note: this basically matches the PRI, I'd like to ask if we can merge this somewhat as-is and make further changes (editorial and otherwise) in new PRs on main.  Note that https://unicode-org.atlassian.net/browse/CLDR-16909 is for changing `techpreview` to `44`, also affecting the spec.

**Comment by miloush** [comment]
As we discussed in the group I would like to have a working implementation before we switch this from preview.

**Comment by srl295** [comment]
> As we discussed in the group I would like to have a working implementation before we switch this from preview.

Keyman is able to support a lot already- still working on regex implementation. So there is a lot that is implemented and also full validation is in the kmc compiler. I just got reordering working recently.

**Comment by mcdurdin** [comment]
I'd be comfortable with waiting for the implementation in Keyman to prove the spec before merging to main. I don't think there is any harm in waiting?

**Comment by srl295** [comment]
@miloush @mcdurdin merging here isn't final release, all i'm proposing here is to merge into in-development CLDR.  This way we go back to regular process for making PR changes. 

CLDR data slush is early August (2023), public alpha Aug 30, public beta Sep 20 , public beta Oct 18. 

1. "waiting for the implementation in Keyman" - do you mean a released implementation?  

2. "not removing techpreview" -  As I said I'm fine to leave "techpreview" on for the moment, perhaps until public beta timeframe. 

-----

To sum up,  what I would like to do with this PR and #3114 is to merge the PRI-level spec into CLDR main.

This would be followed up with PRs to address the feedback (which I am working on collating today).

**Comment by srl295** [comment]
ping… can i get a review on this? I'd like to merge this as-is and then address any changes in subsequent PRs.

File: docs/ldml/tr35-keyboards.md [hunk 1]
```diff
@@ -1076,435 +1294,783 @@ On some architectures, applications may directly interact with keys before they
...
 
-If we mark where the current match position is before and after the transform we see:
+**Implementation Notes**
 
-```
-X | Y Z → X B | Z
-```
+Ideally, markers are implemented entirely out-of-band from the normal text stream. However, implementations _may_ choose to map each marker to a [Unicode private-use character](https://www.unicode.org/glossary/#private_use_character) for use only within the implementation’s processing and temporary storage in the input context.
+
+For example, the first marker encountered could be represented as U+E000, the second by U+E001 and so on.  If a regex processing engine were used, then those PUA characters could be processed through the existing regex processing engine.  `[^\uE000-\uE009]` could be used as an expression to match a character that is not a marker, and `[Ee]\u{E000}` could match `E` or `e` followed by the first marker.
```
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1631]: Given our recent discussions on markers in Keyman, I wonder if this advice may be unhelpful/misleading and perhaps we should just leave it to the implementers to figure out how they want to solve? PUA characters could be used by an experimental keyboard for a new orthography, even if that keyboard is not acceptable for publication into CLDR.

File: docs/ldml/tr35-keyboards.md [hunk 2]
```diff
@@ -1076,435 +1294,783 @@ On some architectures, applications may directly interact with keys before they
...
-> * If the value is positive, then the character is a primary character and is sorted based on the order value as the primary key following a previous base character.
->
-> A character with a zero tertiary value is a primary character and receives a sort key consisting of:
->
-> * Primary weight is the order value
-> * Secondary weight is the index of the character. This may be any value (character index, codepoint index) such that its value is greater than the character before it and less than the character after it.
-> * Tertiary weight is 0.
-> * Quaternary weight is the same as the secondary weight.
+    The hex escaping is case insensitive. The value may not match a surrogate or illegal character.
+    The form
```
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1733]: This looks unfinished

File: docs/ldml/tr35-keyboards.md [hunk 3]
```diff
@@ -1076,435 +1294,783 @@ On some architectures, applications may directly interact with keys before they
...
+- **Character classes**
 
-There is no `@error` attribute.
+    `[abc]` `[^def]` `[a-z]` `[ॲऄ-आइ-ऋ]` `[\u093F-\u0944\u0962\u0963]`
 
-For `@from` attributes with a match string length greater than 1, the sort key information (`@order`, `@tertiary`, `@tertiary_base`, `@prebase`) may consist of a space separated list of values, one for each element matched. The last value is repeated to fill out any missing values. Such a list may not contain more values than there are elements in the `@from` attribute:
+    - supported
+    - no Unicode properties such as `\p{…}`
+    - Warning: Character classes look superficially similar to UnicodeSets as defined in [`unicodeSet`](#element-unicodeSet) elements, but they are different. UnicodeSets must be defined with a `unicodeSet` element, and referenced with the `$[unicodeSet]` notation in transforms. UnicodeSets cannot be used directly in a transform.
```
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1751]: > UnicodeSets cannot be used directly in a transform.

This is slightly confusing -- perhaps we need to say something like:

```suggestion
    - Warning: Character classes look superficially similar to UnicodeSets as defined in [`unicodeSet`](#element-unicodeSet) elements, but they are different. UnicodeSets must be defined with a `unicodeSet` element, and referenced with the `$[unicodeSet]` notation in transforms. UnicodeSets cannot be specified inline in a transform, and can only be used indirectly by reference to the corresponding `unicodeSet` element.
```

File: docs/ldml/tr35-keyboards.md [hunk 4]
```diff
@@ -1076,435 +1294,783 @@ On some architectures, applications may directly interact with keys before they
...
+
+    `?` - equivalent of `{0,1}`
+
+- **Numbered Capture Groups**
+
+    `([abc])([def])` (up to 9 groups)
+
+    These refer to groups captured as a set, and can be referenced with the `$1` through `$9` operators in the `to=` pattern. May not be nested.
+
+- **Nested capture groups**
```
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1770]: Non-capture groups?

File: docs/ldml/tr35-keyboards.md [hunk 5]
```diff
@@ -1076,435 +1294,783 @@ On some architectures, applications may directly interact with keys before they
...
+
+    `((a|b|c)|(d|e|f))`
+
+    **Rationale:** Computational and cognitive complexity.
+
+- **Named capture groups**
+
+    `(?<something>)`
+
+    **Rationale:** Cognitive complexity.
```
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1871]: Implementation complexity?

File: docs/ldml/tr35-keyboards.md [hunk 6]
```diff
@@ -1076,435 +1294,783 @@ On some architectures, applications may directly interact with keys before they
...
+
+Used in the `to=`
+
+- **Literals**
+
+    `$$ \$ \\` = `$ $ \`
+
+- **Entire matched substring**
+
+    `*&`
```
Comment (mcdurdin) [inline docs/ldml/tr35-keyboards.md:1903]: Is this right? not `$0`?

**Review by mcdurdin — APPROVED** [review]
This is a solid cleanup of the spec. I've got some minor nits and a couple of places which need further discussion (new tickets?), but nothing major that should block this from merging I think

File: docs/ldml/tr35-keyboards.md (hunk 6 above)
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1903]: I think you are right. 


```suggestion
    `$0`
```

File: docs/ldml/tr35-keyboards.md [hunk 7]
```diff
@@ -526,209 +626,212 @@ Indicates that:
...
->
-> ![keycap hint](images/keycapHint.png)
+> The `to` attribute may also contain the `\m{…}` syntax to insert a marker. See the definition of [markers](#markers).
+
+_Attribute:_ `transform="no"` (optional)
 
-For example, suppose there are the following keys, their output and one transform:
+> The `transform` attribute is used to define a key that never participates in a transform but its output shows up as part of a transform. This attribute is necessary because two different keys could output the same characters (with different keys or modifier combinations) but only one of them is intended to be a dead-key and participate in a transform. This attribute value must be no if it is present.
+>
+> For example, suppose there are the following keys, their output and one transform:
```
[the remark this answers is not in the document]
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:763]: Perhaps it should just say "does not participate until the next keystroke"

File: docs/ldml/tr35-keyboards.md (hunk 1 above)
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1631]:
```suggestion
Markers should be implemented out-of-band from the normal text stream. Implementations should take care that all normal Unicode codepoints, including private-use area (PUA), are not affected by any scheme for encoding markers in text.
```

File: docs/ldml/tr35-keyboards.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply docs/ldml/tr35-keyboards.md:763]: Yep. And perhaps a note that using markers is another way to solve the problem?

File: docs/ldml/tr35-keyboards.md [hunk 8]
```diff
@@ -526,209 +626,212 @@ Indicates that:
...
+_Attribute:_ `gap="true"` (optional)
+
+> The `gap` attribute indicates that this key does not have any appearance, but represents a "gap" of the specified number of key widths. Can be used with `width` to set a width.
+
+```xml
+<key id="mediumgap" gap="true" width="1.5"/>
+```
 
-_Attribute:_ `longPress="optional"` (optional)
+_Attribute:_ `longPress="a b c"` (optional) (discouraged, see [Accessibility])
```
[the remark this answers is not in the document]
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:713]: I'll remove the 'discouraged' and the link, there's other comments about rewriting the Accessibility section.

File: docs/ldml/tr35-keyboards.md [hunk 9]
```diff
@@ -526,209 +626,212 @@ Indicates that:
...
+_Attribute:_ `longPressDefault` (optional)
 
-> The `transform` attribute is used to define a key that never participates in a transform but its output shows up as part of a transform. This attribute is necessary because two different keys could output the same characters (with different keys or modifier combinations) but only one of them is intended to be a dead-key and participate in a transform. This attribute value must be no if it is present.
+> Indicates which of the `longPress` target characters is the default long-press target, which could be different than the first element. Ignored if not in the `longPress` list. Characters in this attribute can be escaped using the `\u{...}` notation.
+> For example, if the `longPressDefault` is a key whose [display](#Element_displays) appears as `{` an implementation might render the key as follows:
+>
+> ![keycap hint](images/keycapHint.png)
 
-_Attribute:_ `multitap` (optional)
+_Attribute:_ `multiTap` (optional) (discouraged, see [Accessibility])
```
[the remark this answers is not in the document]
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:724]: I'll remove the 'discouraged' and the link, there's other comments about rewriting the Accessibility section.

File: docs/ldml/tr35-keyboards.md [hunk 10]
```diff
@@ -526,209 +626,212 @@ Indicates that:
...
 
-_Attribute:_ `transform="no"` (optional)
+_Attribute:_ `longPressDefault` (optional)
 
-> The `transform` attribute is used to define a key that never participates in a transform but its output shows up as part of a transform. This attribute is necessary because two different keys could output the same characters (with different keys or modifier combinations) but only one of them is intended to be a dead-key and participate in a transform. This attribute value must be no if it is present.
+> Indicates which of the `longPress` target characters is the default long-press target, which could be different than the first element. Ignored if not in the `longPress` list. Characters in this attribute can be escaped using the `\u{...}` notation.
+> For example, if the `longPressDefault` is a key whose [display](#Element_displays) appears as `{` an implementation might render the key as follows:
+>
+> ![keycap hint](images/keycapHint.png)
```
[the remark this answers is not in the document]
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:722]: pre-existing.

File: docs/ldml/tr35-keyboards.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:763]: By the way, a marker that’s never directly consumed, whose purpose may be just to be not-a-character, might need to be noted as far as validation goes.

File: docs/ldml/tr35-keyboards.md (hunk 1 above)
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1631]: This was implemented when this PR merged.

File: docs/ldml/tr35-keyboards.md (hunk 2 above)
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1733]: Fixed in upcoming PR, noting that CLDR-16825 proposes removing \u1234 format

File: docs/ldml/tr35-keyboards.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:724]: * Addressed in upcoming PR

File: docs/ldml/tr35-keyboards.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:713]: * Addressed in upcoming PR

File: docs/ldml/tr35-keyboards.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:763]: adressed in upcoming PR

File: docs/ldml/tr35-keyboards.md (hunk 3 above)
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1751]: Added to upcoming PR

File: docs/ldml/tr35-keyboards.md (hunk 4 above)
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1770]: reworded in upcoming PR

File: docs/ldml/tr35-keyboards.md (hunk 5 above)
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1871]: Yes good catch.!

Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1871]: added to upcoming PR

File: docs/ldml/tr35-keyboards.md (hunk 6 above)
Comment (srl295) [reply docs/ldml/tr35-keyboards.md:1903]: added to upcoming PR
